File: src/leetcode/merge_intervals.cpp
/**
 * Problem link: https://leetcode.com/problems/merge-intervals
 *
 * Given a collection of intervals, merge all overlapping intervals.
 *
 * For example,
 * Given [1,3],[2,6],[8,10],[15,18],
 * return [1,6],[8,10],[15,18].
 */
#include <stdlib.h>

struct Interval {
    int start;
    int end;
};

typedef struct Interval Interval; 
void sort(struct Interval*, struct Interval*);

struct Interval* merge(struct Interval* intervals, int intervalsSize, int* returnSize) {
    sort(intervals, intervals + intervalsSize);
    Interval* result = (Interval*) malloc(intervalsSize * sizeof(Interval));
    int count = 0;
    Interval latest;
    for(int i = 0 ; i < intervalsSize ; ++ i) {
        if(i == 0) {
            latest = intervals[i];
        } else {
            Interval tmp = intervals[i];
            if(tmp.start <= latest.end) {
                if(tmp.end > latest.end) {
                    latest.end = tmp.end;
                }
            } else {
                result[count ++] = latest;
                latest = intervals[i];
            }
        }
    }
    if(intervalsSize > 0) {
        result[count ++] = latest;
    }
    (*returnSize) = count;
    return result;
}

int comp(struct Interval* left, struct Interval* right) {
    return left->start < right->start ? -1 : 1;
}

void sort(struct Interval* start, struct Interval* end) {
    if(end - start <= 1) {
        return;
    }
    struct Interval* pstart = start;
    struct Interval* pend = end - 1;
    Interval key = (*start);
    do {
        while(pend > pstart && comp(pend, &key) > 0) {
            -- pend;
        }

        (*pstart) = (*pend);
        while(pend > pstart && comp(pstart, &key) < 0) {
            ++ pstart;
        }
        (*pend) = (*pstart);
    } while(pend > pstart);

    (*pstart) = key;

    sort(start, pstart);
    sort(pstart + 1, end);
}

int main(int argc, char **argv) {
    return 0;
}

File: src/leetcode/MergeIntervals.java
/**
 * Problem link: https://leetcode.com/problems/merge-intervals
 *
 * Given a collection of intervals, merge all overlapping intervals.
 *
 * For example,
 * Given [1,3],[2,6],[8,10],[15,18],
 * return [1,6],[8,10],[15,18].
 */

import java.util.*;

public class MergeIntervals {
    public static void main(String[] args) {
    }

    public class Interval {
        int start;
        int end;
        Interval() { start = 0; end = 0; }
        Interval(int s, int e) { start = s; end = e; }
    }

    public List<Interval> merge(List<Interval> intervals) {
        Collections.sort(intervals, new IntervalComparator());
        int size = intervals.size();
        Interval latest = null;
        List<Interval> result = new ArrayList<>();
        for(Interval interval : intervals) {
            if(latest == null) {
                latest = interval;
            } else {
                if(interval.start <= latest.end) {
                    if(interval.end > latest.end) {
                        latest.end = interval.end;
                    }
                } else {
                    result.add(latest);
                    latest = interval;
                }
            }
        }

        if(latest != null) {
            result.add(latest);
        }

        return result;
    }

    class IntervalComparator implements Comparator<Interval> {
        public int compare(Interval o1, Interval o2) {
            if(o1.start < o2.start) {
                return -1;
            } else if(o1.start > o2.start) {
                return 1;
            } else {
                if(o1.end < o2.end) {
                    return -1;
                } else if(o1.end > o2.end) {
                    return 1;
                } else {
                    return 0;
                }
            }
        }
    }
}
